traits/trait_types.py

```diff
@@ -3336,6 +3338,124 @@ def as_ctrait(self):
         """
         return self.trait_maker.as_ctrait()
 
+# -------------------------------------------------------------------------------
+#  'None' trait:
+# -------------------------------------------------------------------------------
+
+
+class NoneTrait(TraitType):
+    """ Defines a trait that only accepts the None value
+    """
+
+    info_text = "a None type"
+
+    def validate(self, obj, name, value):
+        if value is None:
+            return value
+
+        self.error(obj, name, value)
+
+# -------------------------------------------------------------------------------
+#  'Union' trait:
+# -------------------------------------------------------------------------------
+
+
+class Union(TraitType):
+    """ Defines a trait whose value can be any of of a specified list of
+    trait types or list of trait type instances or None
+    """
+    # CTrait type map for special trait types:
+    type_map = {"event": TraitKind.event, "constant": TraitKind.constant}
+
+    def __init__(self, *traits, **metadata):
+        self.list_ctrait_instances = []
+        for trait in traits:
+            if trait is None:
+                trait = NoneTrait
+            if not isinstance(trait, (TraitType, type)):
+                raise ValueError("Union trait declaration expects a trait "
+                                 "type or an instance of trait type or None,"
+                                 " but got {} instead".format(trait))
+
+            ctrait_instance = trait().as_ctrait()
+            self.list_ctrait_instances.append(ctrait_instance)
```

Comment:
This might be a good variable to expose via the `inner_traits()` method:
```
def inner_traits(self):
    return self. list_ctrait_instances
```

Reply:
Done